cpp/src/arrow/filesystem/test_util.cc

```diff
@@ -646,6 +647,114 @@ void GenericFileSystemTest::TestGetTargetStatsSelector(FileSystem* fs) {
   ASSERT_RAISES(IOError, fs->GetTargetStats(s, &stats));
 }
 
+using ::testing::ElementsAre;
+
+FileStats File(std::string path) {
+  FileStats st;
+  st.set_type(FileType::File);
+  st.set_path(path);
+  return st;
+}
+
+FileStats Dir(std::string path) {
+  FileStats st;
+  st.set_type(FileType::Directory);
+  st.set_path(path);
+  return st;
+}
+
+// Utility functions to help testing/debugging
+std::ostream& operator<<(std::ostream& os, const FileStats& stats) {
+  return os << "FileStats(" << stats.type() << ", " << stats.path() << ")";
+}
+
+bool operator==(const FileStats& lhs, const FileStats& rhs) {
+  return lhs.type() == rhs.type() && lhs.path() == rhs.path();
+}
+
+void GetSortedStats(FileSystem* fs, Selector s, std::vector<FileStats>& stats) {
+  ASSERT_OK(fs->GetTargetStats(s, &stats));
+  SortStats(&stats);
+}
+
+void GenericFileSystemTest::TestGetTargetStatsSelectorWithRecursion(FileSystem* fs) {
+  ASSERT_OK(fs->CreateDir("01/02/03/04"));
+  ASSERT_OK(fs->CreateDir("AA"));
+  CreateFile(fs, "00.file", "00");
+  CreateFile(fs, "01/01.file", "01");
+  CreateFile(fs, "AA/AA.file", "aa");
+  CreateFile(fs, "01/02/02.file", "02");
+  CreateFile(fs, "01/02/03/03.file", "03");
+  CreateFile(fs, "01/02/03/04/04.file", "04");
+
+  std::vector<FileStats> stats;
+  Selector s;
+
+  s.base_dir = "";
+  s.recursive = false;
+
+  ASSERT_OK(fs->GetTargetStats(s, &stats));
+  SortStats(&stats);
+  EXPECT_THAT(stats, ElementsAre(File("00.file"), Dir("01"), Dir("AA")));
+
+  // recursive should prevail on max_recursion
+  s.max_recursion = 9000;
+  GetSortedStats(fs, s, stats);
+  EXPECT_THAT(stats, ElementsAre(File("00.file"), Dir("01"), Dir("AA")));
+
+  // recursive but no traversal
+  s.recursive = true;
+  s.max_recursion = 0;
+  GetSortedStats(fs, s, stats);
+  EXPECT_THAT(stats, ElementsAre(File("00.file"), Dir("01"), Dir("AA")));
+
+  s.recursive = true;
+  s.max_recursion = 1;
+  GetSortedStats(fs, s, stats);
+  EXPECT_THAT(stats, ElementsAre(File("00.file"), Dir("01"), File("01/01.file"),
+                                 Dir("01/02"), Dir("AA"), File("AA/AA.file")));
+
+  s.recursive = true;
+  s.max_recursion = 2;
+  GetSortedStats(fs, s, stats);
+  EXPECT_THAT(stats, ElementsAre(File("00.file"), Dir("01"), File("01/01.file"),
+                                 Dir("01/02"), File("01/02/02.file"), Dir("01/02/03"),
+                                 Dir("AA"), File("AA/AA.file")));
+
+  s.base_dir = "AA";
```

Comment:
Perhaps cut down a bit on the number of similar tests here? For something like S3 the roundtrips are a bit slow.

Reply:
I think it's better to keep the tests and disable some for s3, if possible

Reply:
I removed 2 redundant checks